File: Data Structures & Algorithms/max-water-container/submission-6.java
class Solution {
    public int maxArea(int[] heights) {
        //l = 0, r = heights.length - 1 
        //h = min(heights[l], heights[r]) -> 1 | 6
        //w = r - l -> 7 | 6
        //area = w * h

        //if(nums[l] <= h) l++
        //else r--

        int l = 0, r = heights.length - 1;
        int maxArea = 0;

        while(l < r) {
            int h = Math.min(heights[l], heights[r]);
            int w = r - l;
            int area = w * h;
            maxArea = Math.max(maxArea, area);

            if(heights[l] <= h) l++;
            else r--;
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-6.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # min k = 1       i = 0 < h = 9
        #                  4 + 3 + 2 + 1 <= h 
        #                  1 + 2 + 2 + 1
        # res = 0
        # [1, 4, 3, 2] -> [0, 0, 0, 0] 
        # 1 -> max(piles) -> monotonic ->+
        # 1 2 3 4 -> m = l + (r - l) // 2
        #   k = mid
        #   totalTime = 0 
        #   for i in range(len(nums)):
        #      totalTime += ceil(nums[i] / k)
        #   if totalTime <= h:
        #      res = k
        #      r = k - 1
        #   else:
        #      l = k + 1    

        res = 0
        l, r = 1, max(piles)
        while l <= r:
            k = l + (r - l) // 2
            totalTime = 0
            for i in range(len(piles)):
                totalTime += math.ceil(piles[i] / k)
            if totalTime <= h:
                res = k
                r = k - 1
            else:
                l = k + 1

        return res